### PR TITLE
Display warning about spaces in underlying values for multiple selection questions. Fixes #1964

### LINKED ISSUE #1964: Display warning when select_multiple names have spaces in them
For questions with multiple selection, the responses chosen by the user are stored as a space-separated list of the response `value`s. Most form builders either warn or error when a user tries to build a form with multiple selection questions that have `value`s with spaces but it's possible to include them when writing XML by hand or using external choices.

A little warning should be displayed when a choice has a space in it. This should look like the error that is shown when an image is missing (look for `R.string.file_missing` for examples).

https://github.com/opendatakit/javarosa/issues/263#issuecomment-369019465

Form: 
[multiple-spaces.xml.txt](https://github.com/opendatakit/collect/files/1776932/multiple-spaces.xml.txt) Notice that if you select some items, save the form, exit, reopen it, no values are selected.

Proposed text: "One or more underlying value is invalid because it has spaces in it."

https://github.com/opendatakit/collect/pull/2024 implements this feature for SelectMultiWidget. We need a more general solution, involving these related classes:
GridMultiWidget
SpinnerMultiWidget
ListMultiWidget
LabelWidget
SelectMultipleAutocompleteWidget
SelectMultiImageMapWidget

### PR DESCRIPTION
**Pull request by bartekfi**
Closes  #1964 

#### What has been done to verify that this works as intended?
Tested with [this form](https://www.dropbox.com/s/ebs4uub4k4qiaxd/multi-warning.xml?dl=0)

     

#### Why is this the best possible solution? Were any other approaches considered?
The solution generalises solution from https://github.com/opendatakit/collect/pull/2024
It adds a TextView in help text layout, and uses it to display a warning.

#### Are there any risks to merging this code? If so, what are they?
The solution can be used to displayed any warning, but is not good for displaying multiple unrelated warnings, as it will always replace previously set warning with a new one.

### CODE REVIEW

**Comment by bartekfi** [comment]
Unrelated to this PR, I've noticed that LabelWidget layout is broken (not affected by this PR).
I also think that enumerating underlying values might not provide any value to app user. 

Instead I would rephrase the warning into something like "Warning: values in this question contain spaces - it might not be saved correctly."

File: collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.Lists;
+import org.javarosa.core.model.SelectChoice;
+import org.odk.collect.android.R;
+import org.odk.collect.android.utilities.UnderlyingValuesConcat;
+import org.odk.collect.android.widgets.QuestionWidget;
+
+import java.util.List;
+
+public class SpacesInUnderlyingValuesWarning {
```
Comment (dcbriccetti) [inline collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java]: A doc comment here would help readers understand what this is

Comment (bartekfi) [reply collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java]: @dcbriccetti Added documentation in c4238ea

**Comment by bartoszfilipowicz** [comment]
CI failed on downloading dependency. Restart needed.

> FAILURE: Build failed with an exception.
> What went wrong:
> Execution failed for task ':collect_app:androidDependencies'.
> Could not resolve all artifacts for configuration ':collect_app:debugCompileClasspath'.
   > Could not find crashlytics-core.aar (com.crashlytics.sdk.android:crashlytics-core:2.6.2).
     Searched in the following locations:
         https://jcenter.bintray.com/com/crashlytics/sdk/android/crashlytics-core/2.6.2/crashlytics-core-2.6.2.aar

File: collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            warningCreator = new WarningTextCreatorImpl();
+        }
+
+        @Override
+        public void render(List<SelectChoice> invalidItems) {
+            questionWidget.showWarning(warningCreator.create(invalidItems, questionWidget.getContext()));
+        }
+    }
+
+    interface SpacesInUnderlyingValues {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java]: It feels too bad that the interface name is so specific. This really could be called something more generic like `UnderlyingValueChecker`, right? I also don't love using `Impl` as part of a class name so perhaps that could then become `SpacesUnderlyingValueChecker` or something like that.

File: collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                throw new IllegalStateException("check() must be called before other methods first");
+            }
+        }
+    }
+
+    public interface WarningTextCreator {
+        String create(List<SelectChoice> invalidValues, Context context);
+    }
+
+    private static class WarningTextCreatorImpl implements WarningTextCreator {
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java]: Instead of `WarningTextCreatorImpl` how about something like `SpacesWarningTextCreator`?

**Review by lognaturel — COMMENTED** [review]
Thank you, @bartoszfilipowicz! I made a couple of small naming suggestions inline.

**Comment by lognaturel** [comment]
> I've noticed that LabelWidget layout is broken

Do you mean in the sense that hint or warning text may overlap with labels? This is actually ok. In actual usage it's rare to have a hint there and it does look good on larger screens where that type of layout is more commonly used. If it's something else, please do file an issue!

> I also think that enumerating underlying values might not provide any value to app user.

I see your point. What I like about including the value(s) is that at least it gives the user something specific they can search for in their form. Users don't always have deep knowledge of form structure and so they may not know what "values in this question" refers to.

File: collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java (hunk 2 above)
Comment (bartekfi) [reply collect_app/src/main/java/org/odk/collect/android/widgets/warnings/SpacesInUnderlyingValuesWarning.java]: Yes, that looks much better, I'll apply the changes.

**Comment by bartekfi** [comment]
@lognaturel  PR updated with propsed name changes. 

> Users don't always have deep knowledge of form structure and so they may not know what "values in this question" refers to.

Understood. Seems like this information is mostly actionable for the form creator, who's testing the form, rather than the person who actually collects the data.

I think this kind of error requires stronger accent, as effectively it makes the data collected for the entire question invalid or incomplete. I propose using red warning colour as in the screenshot below. WDYT @lognaturel?

**Comment by lognaturel** [comment]
> Seems like this information is mostly actionable for the form creator, who's testing the form, rather than the person who actually collects the data.

Exactly!

> I propose using red warning colour as in the screenshot below. 

Thanks, great idea. Once you make that change we're ready for QA unless @dcbriccetti has any other suggestions or questions.

**Comment by dcbriccetti** [comment]
>unless @dcbriccetti has any other suggestions or questions.

Nope

**Review by lognaturel — APPROVED** [review]

**Comment by bartekfi** [comment]
Added one more requested change: changed color of the warning to red. It looks as follows in light and dark themes:

 

This PR is ready for testing!

File: collect_app/src/main/res/layout/help_text_layout.xml [hunk 4]
```diff
@@ -53,5 +53,6 @@
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:visibility="gone"
+            android:textColor="?warningColor"
```
Comment (grzesiek2010) [inline collect_app/src/main/res/layout/help_text_layout.xml]: If it's red in both themes it doesn't really make sense to do that in that way, why don't you just use `@color/red` here?

Comment (bartekfi) [reply collect_app/src/main/res/layout/help_text_layout.xml]: @grzesiek2010 I see your point. My thinking in this matter is as follows:

1. The fact we can use the same color (`@color/red`) for warning in both themes is a coincidence and not a rule or design decision
2. Using attribute allows us to name the color according to its use, rather than value "red". So using `?warningColor` is a nice way to create an alias that is defined by usage not value. (alternatively we can create a global color such as `@color/warningColor`
3. Warning color depends on the theme. It depends on the same looks of theme as `?primaryTextColor`, `?colorAccent`. So theme is the context of the color, and the reason it has this value, so it seems like a natural place to put it. 

WDYT?

Comment (bartekfi) [reply collect_app/src/main/res/layout/help_text_layout.xml]: Using attributes over direct color reference is also recommended by ODK's contribution guidelines.

Comment (lognaturel) [reply collect_app/src/main/res/layout/help_text_layout.xml]: This is not one I feel too strongly about but in general I do prefer semantic naming so I like the approach @bartoszfilipowicz has taken. @grzesiek2010 do you find his arguments convincing?

Comment (grzesiek2010) [reply collect_app/src/main/res/layout/help_text_layout.xml]: > do you find his arguments convincing?

not entirely but I'm not going to complain since it's a small thing. I just noticed the same case is with `colorAccent`, both colors are the same across both themes so adding them to both themes is a duplication I think but fair enough.

**Comment by mmarciniak90** [comment]
@bartoszfilipowicz @lognaturel 
The toast was added to `Multi-select` widget as a part of #2024.
Other widgets do not contain toast, just warning. I feel that we should take care of behavior compatibility for all `multi` widgets.
We need to decide whether add toast for others widgets or remove it from one. 
Removing refers to only one widget so should be faster but on the other hand, the user can be confused what option can cause a problem in the case with images.
screenshot_20180608-152035

**Comment by mmarciniak90** [comment]
@opendatakit-bot label "needs discussion"

**Comment by lognaturel** [comment]
Thanks, @mmarciniak90! I don't think the toast adds much information. Like @bartoszfilipowicz stated previously, the warning is for a form developer rather than for someone collecting data so they should be able to figure out where the changes are needed.

I think we should merge this and then file another issue to remove the toast.

**Comment by mmarciniak90** [comment]
@lognaturel sounds good to me! I created a new issue: #2284

**Tested with success**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Comment by bartekfi** [comment]
Thanks for testing and for reporting issue @mmarciniak90 👍 
@lognaturel I'll post a PR with fix to #2284 as soon as this is merged.
